nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/test/java/org/apache/nifi/json/TestJsonTreeRowRecordReader.java

```diff
@@ -1128,29 +1140,46 @@ void testStartFromNestedFieldThenStartObject() throws IOException, MalformedReco
                 }})
         );
 
-        testReadRecords(jsonPath, expectedRecordSchema, expected, StartingFieldStrategy.NESTED_FIELD, "accounts");
+        testReadRecords(jsonPath, expectedRecordSchema, expected, StartingFieldStrategy.NESTED_FIELD,
+                "accounts", SchemaApplicationStrategy.SELECTED_PART);
+    }
+
+    @Test
+    void testStartsFromNestedObjectWithWholeJsonSchemaApplicationStrategy() throws IOException, MalformedRecordException {
+        String jsonPath = "src/test/resources/json/single-element-nested.json";
+
+        RecordSchema recordSchema = getSchema();
+
+        RecordSchema expectedRecordSchema = getAccountSchema();
+
+        List<Object> expected = Collections.singletonList(
+                new MapRecord(expectedRecordSchema, new HashMap<String, Object>() {{
+                    put("id", 42);
+                    put("balance", 4750.89);
+                }})
+        );
+
+        testReadRecords(jsonPath, recordSchema, expected, StartingFieldStrategy.NESTED_FIELD,
+                "account", SchemaApplicationStrategy.WHOLE_JSON);
+
     }
```

Comment:
Unless the test data creation is really complicated, tests are easier to understand if the data creation is inlined (not done via extracted methods).

Doing that, one can check the input, the metadata (schema in this case) and the expected output at one place.
(Input is actually already extracted to file but I guess we can leave that.)

It also helps avoiding making errors/confusions like in this case: the full schema is created with an "account" field and a "balance" field - even though _there is no_ balance field in the data at the root level.

I would suggest this method done like this:
```java
    @Test
    void testStartsFromNestedObjectWithWholeJsonSchemaScope() throws IOException, MalformedRecordException {
        String jsonPath = "src/test/resources/json/single-element-nested.json";

        RecordSchema accountSchema = new SimpleRecordSchema(Arrays.asList(
            new RecordField("id", RecordFieldType.INT.getDataType()),
            new RecordField("balance", RecordFieldType.DOUBLE.getDataType())
        ));

        RecordSchema recordSchema = new SimpleRecordSchema(Arrays.asList(
            new RecordField("account", RecordFieldType.RECORD.getRecordDataType(accountSchema))
        ));

        RecordSchema expectedRecordSchema = accountSchema;

        List<Object> expected = Collections.singletonList(
                new MapRecord(expectedRecordSchema, new HashMap<String, Object>() {{
                    put("id", 42);
                    put("balance", 4750.89);
                }})
        );

        testReadRecords(jsonPath, recordSchema, expected, StartingFieldStrategy.NESTED_FIELD,
                "account", SchemaScope.WHOLE_JSON);
    }
```
Also I would add a case where the nested field is an array:
```java
    @Test
    void testStartsFromNestedArrayWithWholeJsonSchemaScope() throws IOException, MalformedRecordException {
        String jsonPath = "src/test/resources/json/single-element-nested-array.json";

        RecordSchema accountSchema = new SimpleRecordSchema(Arrays.asList(
            new RecordField("id", RecordFieldType.INT.getDataType()),
            new RecordField("balance", RecordFieldType.DOUBLE.getDataType())
        ));

        RecordSchema recordSchema = new SimpleRecordSchema(Arrays.asList(
            new RecordField("accounts", RecordFieldType.ARRAY.getArrayDataType(RecordFieldType.RECORD.getRecordDataType(accountSchema)))
        ));

        RecordSchema expectedRecordSchema = accountSchema;

        List<Object> expected = Arrays.asList(
                new MapRecord(expectedRecordSchema, new HashMap<String, Object>() {{
                    put("id", 42);
                    put("balance", 4750.89);
                }}),
                new MapRecord(expectedRecordSchema, new HashMap<String, Object>() {{
                    put("id", 43);
                    put("balance", 48212.38);
                }})
        );

        testReadRecords(jsonPath, recordSchema, expected, StartingFieldStrategy.NESTED_FIELD,
                "accounts", SchemaScope.WHOLE_JSON);
    }
```

Reply:
Thanks!